spec.html

```diff
@@ -32609,68 +32609,31 @@ <h1>Array.prototype.some ( _callbackfn_ [ , _thisArg_ ] )</h1>
-        <p>If _comparefn_ is not *undefined* and is not a consistent comparison function for the elements of this array (see below), the sort order is implementation-defined. The sort order is also implementation-defined if _comparefn_ is *undefined* and SortCompare does not act as a consistent comparison function.</p>
-        <p>Let _proto_ be _obj_.[[GetPrototypeOf]](). If _proto_ is not *null* and there exists an integer _j_ such that all of the conditions below are satisfied then the sort order is implementation-defined:</p>
-        <ul>
-          <li>
-            _obj_ is sparse
-          </li>
-          <li>
-            0 &le; _j_ &lt; _len_
-          </li>
-          <li>
-            HasProperty(_proto_, ToString(_j_)) is *true*.
-          </li>
-        </ul>
-        <p>The sort order is also implementation-defined if _obj_ is sparse and any of the following conditions are true:</p>
-        <ul>
-          <li>
-            IsExtensible(_obj_) is *false*.
-          </li>
-          <li>
-            Any integer index property of _obj_ whose name is a nonnegative integer less than _len_ is a data property whose [[Configurable]] attribute is *false*.
-          </li>
-        </ul>
-        <p>The sort order is also implementation-defined if any of the following conditions are true:</p>
-        <ul>
-          <li>
-            If _obj_ is an exotic object (including Proxy exotic objects) whose behaviour for [[Get]], [[Set]], [[Delete]], and [[GetOwnProperty]] is not the ordinary object implementation of these internal methods.
-          </li>
-          <li>
-            If any index property of _obj_ whose name is a nonnegative integer less than _len_ is an accessor property or is a data property whose [[Writable]] attribute is *false*.
-          </li>
-          <li>
-            If _comparefn_ is *undefined* and the application of ToString to any value passed as an argument to SortCompare modifies _obj_ or any object on _obj_'s prototype chain.
-          </li>
-          <li>
-            If _comparefn_ is *undefined* and all applications of ToString, to any specific value passed as an argument to SortCompare, do not produce the same result.
-          </li>
-        </ul>
-        <p>The following steps are taken:</p>
-        <emu-alg>
+          1. Let _objIsSparse_ be ObjectIsSparse(_obj_, _len_).
+          1. Let _proto_ be _obj_.[[GetPrototypeOf]]().
+          1. Let _sortOrderIsImplementationDefined_ be *true* if any of the following conditions are satisfied, and let it be *false* otherwise:
+            * _objIsSparse_ is *true* and _proto_ is not *null* and there exists a nonnegative integer _j_ less than _len_ such that HasProperty(_proto_, ToString(_j_)) is *true*.
```

Comment:
```suggestion
            * _objIsSparse_ is *true* and _proto_ is not *null* and there exists a nonnegative integer _j_ less than _len_ such that ! HasProperty(_proto_, ! ToString(_j_)) is *true*.
```

Reply:
Yes to the `!` for `ToString`. As for `HasProperty`, I think that has to be `?`.

But now I'm wondering: with the status quo, if `HasProperty` (or `IsExtensible`, or various others) returns an abrupt completion, does that cause `sort` to return an abrupt completion, or does that merely cause the condition in question to not be satisfied, which then only affects whether the sort order is implementation-defined? (Note that these conditions are outside the "Perform a sequence of calls" step, so its point about abrupt completions doesn't apply.)